src/events/models.py

```diff
@@ -49,30 +49,173 @@ def get_probes(self):
+    """
+    def __init__(self, start_time, stop_time, probes,
+                 threshold_interval, threshold_p):
+        self.start_time = start_time
+        self.stop_time = stop_time
+        self.probe_ids = [p.id for p in list(probes)]
+        self.threshold_interval = relativedelta(seconds=threshold_interval)
+        self.threshold_p = threshold_p
+
+    def _get_connect_and_disconnect_events(self):
+        """
+        Return all of the connect and disconnect events from Atlas.
+
+        These events are a list of dict, each of which looks like 
+        this:
+
+            {'asn': '3292',
+             'controller': 'ctr-nue16',
+             'event': 'connect',
+             'msm_id': 7000,
+             'prb_id': 22964,
+             'prefix': '83.88.0.0/13',
+             'timestamp': 1463848675,
+             'type': 'connection'},
+
+        This list is sorted by timestamp.
+        """
+        results = AtlasResultsRequest(
+            # 7000 is the undocumented magical measurement ID which 
+            # returns the connect & disconnect events
+            msm_id = 7000,
+            start = self.start_time,
+            stop = self.stop_time,
+            probe_ids = self.probe_ids
         ).create()[1]
+        return sorted(results, key=lambda x: x['timestamp'])
 
     def get(self):
-        pass
+        """
+        Look through our disconnect events and find when we have a
+        noteworthy outage event.
+
+        We can figure out the number of probes that indicates this.
+        That is the (total number of probes) * (alert percentage).
+
+        We then look for any time period that has that many disconnect
+        events from different probes. We use an inefficient algorithm
+        for this, because there are typically only a few such events.
+        (An hour time slice in all of Denmark had 8 in a quick test.)
+
+        Returns a sorted list of time ranges, along with a set of
+        probes failing during each range. The list covers the whole
+        time period, so the set of probes failing for any given time
+        may be empty. Like this:
+
+        [ 
+          { 'start_time': 123,
+            'stop_time': 456, 
+            'probe_ids': set([555, 666, 777]) },
+          { 'start_time': 457,
+            'stop_time': 999,
+            'probe_ids': set() },
+          { 'start_time': 1000,
+            'stop_time': 1001,
+            'probe_ids': set([111, 555, 666, 777] },
+        ]
+        """
+        failed_level = int(len(self.probe_ids) * self.threshold_p)
+
+        # search through our events and find disconnect events 
+        # that are within the time period
+        all_evts = self._get_connect_and_disconnect_events()
+        unpacked_outages = []
+        for n, evt in enumerate(all_evts):
+            # use a set to insure that we only track each probe
+            # that disconnects once
+            down_prb_ids = set([evt['prb_id']])
```

Comment:
PyCharm is telling me that this should be a set literal (rather than calling `set()` but I'm not sure if this should be `{evt["prb_id"]}` or `{[evt["prb_id"]]}`.